pyk/src/pyk/proof/reachability.py

```diff
@@ -728,6 +738,7 @@ def _inject_module(module_name: str, import_name: str, sentences: list[KRule]) -
             for rule in subproof.as_rules(priority=20, direct_rule=self.direct_subproof_rules)
         ]
         circularity_rule = proof.as_rule(priority=20)
+        self.circularity_rule_id = circularity_rule.label
```

Comment:
During the lifetime of an `APRProver`, is `init_proof` allowed to be called multiple times with different proofs? So for example, is this a valid interaction?

```python3
prover.init_proof(a)       # write circularity_rule_id
prover.init_proof(b)       # write circularity_rule_id
...
prover.step_proof(a_step)  # read circularity_rule_id
```

If so, then `circularity_rule_id` would be better as an attribute on `APRProof`, `APRProofStep`, etc.

Reply:
`Proof.init_proof` is called at the beginning of `Proof.advance_proof`, and should be called before doing something like `Proof.step_proof`. So I guess this pattern would not be allowed, but also is not prevented.

The issue is that `Proof.step_proof` does not have access to the `Proof`, but a `ProofStep`, so we'd need to pass that information in addition on teh `ProofStep`. I can do that if you'd prefer, as a `str | None` basically.

Reply:
Method `step_proof` comes from the abstract superclass, so if we add a parameter, it'll have to be added for all classes that implement it. And `circularity_rule_id` looks specific to `APRProof`.

Would it make sense to make `circularity_rule_id` an attribute of `APRProof`? Then `APRProof.get_steps` can pass that value to `APRProofStep`, so it becomes accessible in `APRProver.step_proof`.

If not, the lifecycle of `APRProver` can still be documented and invalid interactions checked during runtime (as long as it's not inconsistent with the abstract `Prover` lifecycle, e.g. what `parallel_advance_proof` expects from the interface).

@nwatson22, what's your take on this?

Reply:
We don't want to store a `circularity_rule_id` in the prover, since it's proof-specific, like you say. I think adding it to `APRProof` makes sense. Another similar option might be if we could assign the circularity rule a specific label when we create it that we would be able to calculate just from the proof ID, and then we wouldn't have to explicitly include it in the proof step.

Reply:
Ok, I've refactored it so that the `APRProof` computes/carries the rule id, and passes it into the `APRProver` via the `APRProofStep`.